Refactor to allow more flavours of test to be automatically collected

duog opened the pull request:
Hi,

I've been playing with tasty-th over the last couple of days. I was frustrated by not being able to collect HSpec tests. I have made some changes to add the following features:
1. Tests of type IO TestTree can now  be collected. This allows Hspec tests, and presumably golden tests, to be automatically collected.
2. The type signature of test_ top-level definitions is relaxed, whereas they were required to be [TestTree], now TestTree, [TestTree], IO TestTree, IO [TestTree], [IO TestTree] are all acceptable.
3. Each of the main template haskell functions has an additional variant that accepts arguments, which can be used to customize and add new varieties of tests.
4. I've added a test suite that exercises the majority of the library

Note that some uses have been broken, details below. 

Please let me know if you are interested in merging and I will tidy up this pull request with haddocks, exports, and any other changes you need.

The breakage is in the type of the expression generated by the $(testGroupGenerator*) splices. This type is now IO TestTree, where it was TestTree.
I can think of the following uses that will break:
1. 
`testgroup :: TestTree`
`testgroup = $(testGroupGenerator)` 
This can be fixed by removing the type annotation or changing it to IO TestTree

2.
'atesttree :: TestTree`
`prop_aproperty :: Integer -> Bool`
`test_testgroupsfromotherfiles = [testgroupfromfile1, atesttree ]`
This fails because testgroupfromfile1 is an IO TestTree. It can be fixed by writing`return atesttree` in the list. I don't know how many people mix TestTrees and splices like this though.

duog committed:
Added .stack-work to .gitignore
Added stack.yaml

duog committed:
Introduced TestVarieties to model different flavours of test. The data type TestVariety consists of a predicate to determine whether a variety supports a testname, and a template haskell splice to generate a function call that will take a name and return an IO TestTree
Refactored TH.hs to use TestVarieties, the interface should have remained the same. There are additional functions - such as defaultMain' - that take Args as a parameter, these have not been exported yet.
Added a test suite to exercise tasty-th.

duog committed:
Moved Template haskell to the top of the files to demonstrate that tests before the template haskell is not required.

duog committed:
Add check that Test.Tasty.QuickCheck.testProperty is in scope, with a template haskell error if it is not. I intend to extend this to consider also QC.testProperty, TQC.testProperty and testProperty, and to apply the same technique to the other varieties. Also need to clean up the error message.

duog committed:
refactor testGroupGeneratorFor' to use TastyGroupable @done(2016-02-28 22:55)
Refactor varieties: predicate out of TestVariety and move name into a parameter for the splice @done(2016-02-29 14:10)

duog committed:
Generalize TastyGroupable instances to use Foldable/Travsersable instead of lists.

duog committed:
Reworked TastyGroupable instances
Added test for all TastyGroupable cases and spec_

duog committed:
Delete TaskList.todo

bennofs commented:
Thanks for your work, this looks great! I would like to merge this after you add the docs/cleanup (I can help with the travis failures, don't worry about that), I've got one question though:

Would it be possible to retain the old type of `$(testGroupGenerator)`, and add a new function `$(testGroupGeneratorIO)` that supports `IO`? I think it's common to use `testGroupGenerator` in a few modules, and then arrange those `TestTrees` in a test tree manually in a `Main` module. Having to pull them all out of IO first makes this harder (it requires some calls to `sequence` and all test grouping functions need to be `fmap`ed and cannot be called directly).

duog commented:
I think that in most cases the toTestTree function on the TastyGroupable type class should be sufficient  to get everything into IO e.g.
`manual_testtree :: TestTree`
`manual_testtree = testGroup "name" [mytest1, $(testGroupGenerator) ...]`

can be changed to 
`manual_testtree :: IO TestTree`
`manual_testtree = toTestTree "name" [toTestTree "" mytest1, $(testGroupGenerator ...)]`

Having said that, it's obviously better to be backwards compatible; I will add testGroupGeneratorIO as you suggest, and I think I can get testGroupGenerator to emit a warning if it sees a spec_ function which it can't deal with.

I'm pretty sure defaultMain delegating to testGroupGeneratorIO will be fine.

Will work on this a bit later and get back to you.

duog committed:
Refactor errors produced by template haskell into an ADT
Add TestTreeSpliceMode which is used by TestVarieties to determine whether to produce expressions of type TestTree or IO TestTree
Add functions testGroupGeneratorIO, testGroupGeneratorIO', testGroupGeneratorForIO, testGroupGeneratorForIO'. These functions produce expressions of type IO TestTree, while the existing functions produce expressions of type TestTree.
Add many exports to TH.hs
Add many tests

samestep commented:
Is there anything else that needs to be changed about this PR before it is merged?
